### ISSUE TITLE
Defer Scroll view gesture to scroll content buttons when sliver does not actually scroll

### ISSUE THREAD

**Issue by Evoswork** [flutter/flutter#90809]
I've found a problem with the event of pushing the button in SliverAppBar. The parameters are: floating: true, snap: true, pinned: false. As I see, while the animation of Sliverlist items scrolling takes place, clicks can't be detected. I mean when you click SliverAppBar it just hides again.

Also I've tried not to hide SliverAppBar, as a result scrolling stops on the first click, and only after the second one the buttons start working
https://user-images.githubusercontent.com/66563284/134918248-76cde7bf-6acf-4562-bda7-3f1fd15fc2a9.mov

    return Scaffold(
      body: SafeArea(
        child: CustomScrollView(
          slivers: [
            SliverAppBar(
              leading: IconButton(
                  onPressed: () {
                    print('click');
                  },
                  icon: const Icon(Icons.arrow_back)),
              floating: true,
              snap: true,
              toolbarHeight: 80,
              backgroundColor: Colors.blue,
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (context, index) => Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Container(
                    width: double.infinity,
                    height: 150,
                    color: Colors.red,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );

/Users/evgenii/flutter/bin/flutter doctor --verbose
[✓] Flutter (Channel stable, 2.5.1, on macOS 11.6 20G165 darwin-x64, locale ru-RU)
    • Flutter version 2.5.1 at /Users/evgenii/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (10 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/evgenii/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/evgenii/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.0, Build version 13A233
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.2)
    • IntelliJ at /Users/evgenii/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5284.31

[✓] IntelliJ IDEA Community Edition (version 2021.2.2)
    • IntelliJ at /Users/evgenii/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/212.5284.40/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Users/evgenii/Library/Application Support/JetBrains/Toolbox/apps/IDEA-C/ch-0/212.5080.55/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.2)
    • IntelliJ at /Users/evgenii/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/212.5284.40/IntelliJ IDEA.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 212.3116.29

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.2)
    • IntelliJ at /Users/evgenii/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/212.5284.19/IntelliJ IDEA.app
    • Flutter plugin version 56.0.5
    • Dart plugin version 212.3116.29

[✓] Connected device (3 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • macOS (desktop)         • macos         • darwin-x64     • macOS 11.6 20G165 darwin-x64
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 94.0.4606.61

• No issues found!
Process finished with exit code 0

**Comment by darshankawar** [comment]
@Evoswork 
Thanks for the report. I ran the sample code and see below behavior:


https://user-images.githubusercontent.com/67046386/135035710-2c8a7a5a-92dc-44ae-ba64-00691479813c.mov


When I tap on back arrow, it seems to recognize the click:

```
flutter: click
flutter: click
flutter: click
flutter: click
flutter: click
flutter: click
```

What's the expected behavior you are looking in this case ?

**Comment by Evoswork** [comment]
You need to click before the scroll

**Comment by darshankawar** [comment]
Thanks for the update. I clicked on back arrow before scrolling, and it recognizes the click.

**Comment by Evoswork** [comment]
I'm sorry, I wrote it wrong.
While scrolling.

**Comment by darshankawar** [comment]
I see what you are saying. I observed same behavior now. When we scroll up, the SliverAppBar shows up which has a back arrow button, but if we try to tap on it, it goes away instead of recognizing the tap event. The gesture is recognized only when the scrolling has stopped, sliverappbar is visible:

flutter: click
flutter: click


https://user-images.githubusercontent.com/67046386/135210161-70ff4647-fda8-4535-8afd-da995cc94e8d.mov

<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 2.5.1, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.5.1 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (2 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!

[✓] Flutter (Channel master, 2.6.0-12.0.pre.134, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.6.0-12.0.pre.134 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0167d99fc6 (5 hours ago), 2021-09-28 17:23:06 -0700
    • Engine revision bccb3a57eb
    • Dart version 2.15.0 (build 2.15.0-156.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!




```

</details>

**Comment by HansMuller** [comment]
CC @Piinks

**Comment by jjoelson** [comment]
This may have the same root cause as this one: https://github.com/flutter/flutter/issues/92119

It seems that no touches within a `SliverAppBar` are recognized while scrolling. This includes:
- Back/dismiss buttons
- Action buttons
- Custom gesture detectors in the title widget
- iOS Status bar taps (for the tap to scroll up gesture)

My guess is that the logic that prevents, say, a button within a scroll view from getting pressed if the user taps during scrolling is not distinguishing between slivers that are moving and slivers that are fixed in place.

**Comment by Piinks** [comment]
> - iOS Status bar taps (for the tap to scroll up gesture)

This case is implemented in the Scaffold rather than the Sliver/AppBar.

I think this is WAI. I added a notification listener to observe the start and end of scrolling as it relates to the button clicks, and the click is being received while scrolling. The timing is tricky because you have float&snap set in your code sample, so the app bar could slide right out from under your finger before you had a chance to tap it. I had more success hovering the mouse pointer over the button and using a mouse scroll to initiate scrolling and click the button, rather than dragging and trying to quickly hit the button.
I think what is a little confusing may be that when you tap on the button, scrolling immediately stops. This is because the sliver app bar is actually part of the scrolling content of the CustomScrollView, so when you tap on it, it puts the brakes on the scrolling content that is rolling by. If there were an AppBar above the CustomScrollView, the behavior would be slightly different because the AppBar is not part of the Scrollable. I think this is expected but if you have more insight, please do share. I have updated the code sample below:

<details>

<summary>Updated Code Sample </summary>

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: NotificationListener<ScrollNotification>(
        onNotification: (notification) {
          print(notification.runtimeType);
          return true;
        },
        child: const Scaffold(
          body: MyBody(),
        ),
      ),
    );
  }
}

class MyBody extends StatelessWidget {
  const MyBody({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return CustomScrollView(
      slivers: [
        SliverAppBar(
          leading: IconButton(
            onPressed: () {
              print('Click');
            },
            icon: const Icon(Icons.arrow_back),
          ),
          // pinned: true,
          floating: true,
          snap: true,
        ),
        SliverList(
          delegate: SliverChildBuilderDelegate(
            (context, index) => Padding(
              padding: const EdgeInsets.all(8.0),
              child: Container(
                child: Center(child: Text('Item $index')),
                height: 150,
                color: Colors.red,
              ),
            ),
          ),
        ),
      ],
    );
  }
}

```

</details>

<details>

<summary>Flutter doctor</summary>

```
[✓] Flutter (Channel master, 2.6.0-12.0.pre.877, on macOS 12.0.1 21A559 darwin-x64, locale en)
    • Flutter version 2.6.0-12.0.pre.877 at /Users/katelovett/_github/flutter
    • Upstream repository git@github.com:flutter/flutter.git
    • Framework revision 0eeb026e58 (54 minutes ago), 2021-12-01 12:19:05 -0800
    • Engine revision c2fcadef89
    • Dart version 2.16.0 (build 2.16.0-56.0.dev)
    • DevTools version 2.8.0

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0-rc2)
    • Android SDK at /Users/katelovett/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0-rc2
    • Java binary at: /Applications/Android Studio with
      Blaze.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio with Blaze.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2.3)
    • IntelliJ at /Applications/IntelliJ UE (stable).app
    • Flutter plugin version 62.0.3
    • Dart plugin version 212.5632

[✓] VS Code (version 1.62.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 12.0.1 21A559 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 96.0.4664.55

! Doctor found issues in 1 category.
```

</details>

**Comment by jjoelson** [comment]
Hi @Piinks! If you set `pinned: true` in the `SliverAppBar` then the issue is far easier to reproduce, and it is also much more obvious why the behavior of stopping scrolling is probably unexpected by users. A `CustomScrollView` with a pinned `AppBarSliver` looks to a user exactly like a screen with a normal `AppBar`. That the back button can be tapped during scrolling with a normal app bar, but not with a pinned `AppBarSliver` is a bit confusing. 

I think many developers have started defaulting to always using `SliverAppBar` on screens with scrolling content because you never know when a designer/PM is going to decide that they actually want that app bar to scroll off-screen, and I think it would be good for this to work the same as a normal `AppBar` when it appears fixed at the top of the screen (either due to `pinned` or `floating`). 

There may not be an easy solution for this, because stopping scrolling on tap is the proper behavior _in general_ for things in scroll views; but certain types of slivers are designed to _not_ scroll despite being within a scroll view, and so I think it's worth a bit of brainstorming to see if it's possible to pass touch events to those slivers even during scrolling.

**Comment by Piinks** [comment]
Thanks @jjoelson! I understand what you mean, well said. I have updated the bug to reflect this, feel free to update further if the title doesn't quite reflect it. 
Thanks for helping to investigate this!

**Comment by darshankawar** [comment]
removing `wait` label per above comment.

**Comment by Piinks** [comment]
I am going to de-dupe this with https://github.com/flutter/flutter/issues/42588 to consolidate. Please follow that issue for updates, thank you!
